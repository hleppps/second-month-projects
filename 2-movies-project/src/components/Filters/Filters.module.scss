.filters {
  border: 1px solid #dadee2;
  padding: 0.5rem;
  border-radius: 0.3rem;
  user-select: none;
}
