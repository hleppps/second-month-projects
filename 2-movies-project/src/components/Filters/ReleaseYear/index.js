import ReleaseYear from "./ReleaseYear";

export default ReleaseYear;
