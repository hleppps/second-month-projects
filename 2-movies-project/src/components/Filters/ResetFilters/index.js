import ResetFilters from "./ResetFilters";

export default ResetFilters;
