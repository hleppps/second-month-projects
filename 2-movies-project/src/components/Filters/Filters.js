import React, { useCallback, useMemo } from "react";
import styles from "./Filters.module.scss";
import Genres from "./Genres";
import Pagination from "./Pagination";
import ReleaseYear from "./ReleaseYear";
import ResetFilters from "./ResetFilters";
import SortBy from "./SortBy";
import * as CONSTANTS from "../../utils/constants";

const Filters = React.memo(
  ({ totalPages, defaultFiltersState, setFilters, filters, ...props }) => {
    // Functions to modify app.js filters state

    const setPage = useCallback(
      (changeType, prevPage) => {
        let newPage;
        if (changeType === CONSTANTS.PAGINATION_INCREASE) {
          newPage = prevPage === totalPages ? totalPages : ++prevPage;
        }

        if (changeType === CONSTANTS.PAGINATION_DECREASE) {
          newPage = prevPage === 1 ? 1 : --prevPage;
        }

        return { page: newPage };
      },
      [totalPages]
    );

    const setYear = useCallback((year) => {
      return { year, page: 1 };
    }, []);

    const setSorting = useCallback((sorting) => {
      return { sorting, page: 1 };
    }, []);

    const setGenres = useCallback((genres) => {
      return { genres, page: 1 };
    }, []);

    const resetFilters = useCallback(() => {
      return defaultFiltersState;
    }, [defaultFiltersState]);

    // Selecting the desired function

    const setFunctions = useMemo(() => {
      return {
        PAGE: setPage,
        YEAR: setYear,
        SORT: setSorting,
        GENRES: setGenres,
        RESET: resetFilters,
      };
    }, [resetFilters, setGenres, setSorting, setYear, setPage]);

    const setFiltersHandler = useCallback(
      (filterType, ...args) => {
        const propertiesToFilter = setFunctions[filterType](...args);

        setFilters((prevState) => {
          return { ...prevState, ...propertiesToFilter };
        });
      },
      [setFilters, setFunctions]
    );

    // Filter handlers with preset types

    const setGenresFilterHandler = useCallback(
      (selectedGenres) => {
        setFiltersHandler("GENRES", selectedGenres);
      },
      [setFiltersHandler]
    );

    const setSortingFilterHandler = (sortingValue) => {
      setFiltersHandler("SORT", sortingValue);
    };

    const setYearFilterHandler = (releaseYear) => {
      setFiltersHandler("YEAR", releaseYear);
    };

    const setPageFilterHandler = (changeType, curPage) => {
      setFiltersHandler("PAGE", changeType, curPage);
    };

    const resetFiltersHandler = () => {
      setFiltersHandler("RESET");
    };

    return (
      <div data-testid="filters" className={styles.filters}>
        <h2>Filters:</h2>

        <>
          <ResetFilters resetFilters={resetFiltersHandler} />
          <SortBy
            setSortingFilter={setSortingFilterHandler}
            filteredSorting={filters.sorting}
          />
          <ReleaseYear
            setYearFilter={setYearFilterHandler}
            filteredYear={filters.year}
          />
          <Genres
            allGenresList={props.allGenresList.genres}
            selectedGenresList={filters.genres}
            setGenresFilter={setGenresFilterHandler}
          />
          <Pagination
            totalPages={totalPages}
            setPageFilter={setPageFilterHandler}
            filteredPage={filters.page}
          />
        </>
      </div>
    );
  }
);

export default Filters;
