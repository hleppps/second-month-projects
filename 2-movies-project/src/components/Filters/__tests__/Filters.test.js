import { cleanup, render } from "@testing-library/react";
import React from "react";
import Filters from "../Filters";

const filters = {
  page: 1,
  sorting: "popularity.desc",
  year: 2021,
  genres: [],
};
const allGenresList = [];

afterEach(cleanup);

describe("Filters components", () => {
  it("renders with content", () => {
    const { getByTestId } = render(
      <Filters filters={filters} allGenresList={allGenresList} />
    );

    expect(getByTestId("filters")).not.toBeNull();
  });

  it("matches a snapshot", () => {
    const { asFragment } = render(
      <Filters filters={filters} allGenresList={allGenresList} />
    );
    expect(asFragment(<Filters />)).toMatchSnapshot();
  });
});
