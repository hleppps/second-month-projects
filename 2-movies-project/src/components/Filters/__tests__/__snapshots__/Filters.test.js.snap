// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filters components matches a snapshot 1`] = `
<DocumentFragment>
  <div
    class="filters"
    data-testid="filters"
  >
    <h2>
      Filters:
    </h2>
    <button
      class="w-100 btn btn-outline-secondary"
      data-testid="reset-button"
      type="button"
    >
      Reset Filters
    </button>
    <div
      class="my-3"
    >
      <label
        class="form-label"
      >
        Sort by:
      </label>
      <select
        aria-label="Default select example"
        class="form-select"
        data-testid="sort-select"
      >
        <option
          value="popularity.desc"
        >
          Most Popular
        </option>
        <option
          value="popularity.asc"
        >
          Least Popular
        </option>
        <option
          value="vote_average.desc"
        >
          Most Wanted
        </option>
        <option
          value="vote_average.asc"
        >
          Least Wanted
        </option>
      </select>
    </div>
    <div
      class="my-3"
    >
      <label
        class="form-label"
      >
        Release year:
      </label>
      <select
        aria-label="Default select example"
        class="form-select"
        data-testid="year-select"
      >
        <option
          value="2021"
        >
          2021
        </option>
        <option
          value="2020"
        >
          2020
        </option>
        <option
          value="2019"
        >
          2019
        </option>
        <option
          value="2018"
        >
          2018
        </option>
      </select>
    </div>
    <button
      class="my-2 w-100 btn btn-outline-secondary"
      data-testid="reset-genres"
      type="button"
    >
      Show All Genres
    </button>
    <div
      class="my-2"
      data-testid="genres-list"
    >
      <div
        class="spinner-grow"
      />
    </div>
    <div
      class="pagination"
    >
      <div
        class="buttons"
      >
        <button
          class="m-1 w-100 btn btn-outline-secondary"
          data-testid="decrease-button"
          type="button"
        >
          «
        </button>
        <button
          class="m-1 w-100 btn btn-outline-secondary"
          data-testid="increase-button"
          type="button"
        >
          »
        </button>
      </div>
      <p
        class="pages"
      >
        <span>
          Page:
        </span>
         1 of 
      </p>
    </div>
  </div>
</DocumentFragment>
`;
